test/e2e/group_setting_test.go

```diff
@@ -0,0 +1,419 @@
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gstruct"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	externaldnsendpoint "sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+	"github.com/kuadrant/dns-operator/internal/common/hash"
+	. "github.com/kuadrant/dns-operator/test/e2e/helpers"
+)
+
+const (
+	controllerNamespace      = "dns-operator-system"
+	controllerDeploymentName = "dns-operator-controller-manager"
+	deploymentReadyTimeout   = 3 * time.Minute
+)
+
+// Group tests are Serial and Ordered to minimize deployment restarts
+// This reduces test time from ~5 minutes to ~2-3 minutes
+var _ = Describe("Group Setting on Controller", Labels{"group"}, Serial, Ordered, func() {
+	var k8sClient client.Client
+	var testDNSProviderSecret *v1.Secret
+	var initialGroupValue string
+
+	BeforeAll(func(ctx SpecContext) {
+		k8sClient = testClusters[0].k8sClient
+		testDNSProviderSecret = testClusters[0].testDNSProviderSecrets[0]
+
+		// Store the initial group value from the deployment
+		deployment := &appsv1.Deployment{}
+		err := k8sClient.Get(ctx, client.ObjectKey{
+			Namespace: controllerNamespace,
+			Name:      controllerDeploymentName,
+		}, deployment)
+		Expect(err).NotTo(HaveOccurred())
+		initialGroupValue = getGroupFromDeployment(deployment)
+	})
+
+	AfterAll(func(ctx SpecContext) {
+		// Restore initial group value only once at the end
+		By("restoring initial controller group configuration")
+		err := setControllerGroup(ctx, k8sClient, initialGroupValue)
+		Expect(err).NotTo(HaveOccurred())
+		waitForDeploymentReady(ctx, k8sClient, controllerNamespace, controllerDeploymentName)
+	})
+
+	// Test 1: Start with no group, create a DNSRecord
+	It("should create DNSRecord without group when controller has no group configured", func(ctx SpecContext) {
+		By("ensuring controller has no group configured")
+		err := setControllerGroup(ctx, k8sClient, "")
+		Expect(err).NotTo(HaveOccurred())
+
+		By("waiting for deployment to become ready")
+		waitForDeploymentReady(ctx, k8sClient, controllerNamespace, controllerDeploymentName)
+
+		By("creating a DNSRecord without group")
+		dnsRecord := createTestDNSRecord(ctx, k8sClient, testDNSProviderSecret, "127.0.0.1")
+
+		By("checking DNSRecord becomes ready without group")
+		Eventually(func(g Gomega, ctx context.Context) {
+			err := k8sClient.Get(ctx, client.ObjectKeyFromObject(dnsRecord), dnsRecord)
+			g.Expect(err).NotTo(HaveOccurred())
+			g.Expect(dnsRecord.Status.Conditions).To(
+				ContainElement(MatchFields(IgnoreExtras, Fields{
+					"Type":   Equal(string(v1alpha1.ConditionTypeReady)),
+					"Status": Equal(metav1.ConditionTrue),
+				})),
+			)
+			g.Expect(string(dnsRecord.Status.Group)).To(BeEmpty())
+		}, recordsReadyMaxDuration, 10*time.Second, ctx).Should(Succeed())
+
+		DeferCleanup(func(ctx SpecContext) {
```

Comment:
I think this DeferCleanup should be earlier in the test, what's the point of deferring it, when it's the last thing done anyway. And won't cleanup not run if the tests fail before this defer is defined?